recipes/llvm-core/all/conanfile.py

```diff
@@ -1,429 +1,496 @@
+        if self.options.get_safe("with_libedit"):
+            self.requires("editline/3.1")
+        if self.options.with_zlib:
+            self.requires("zlib/[>=1.2.11 <2]")
+        if self.options.with_xml2:
+            self.requires("libxml2/[>=2.12.5 <3]")
+        if self.options.with_z3:
+            self.requires("z3/4.13.0")
+
+    def build_requirements(self):
+        self.tool_requires("ninja/1.12.1")
 
     def validate(self):
-        if self.options.shared:  # Shared builds disabled just due to the CI
-            message = 'Shared builds not currently supported'
-            raise ConanInvalidConfiguration(message)
-            # del self.options.fPIC
-        # if self.settings.os == 'Windows' and self.options.shared:
-        #     message = 'Shared builds not supported on Windows'
-        #     raise ConanInvalidConfiguration(message)
+        if self.settings.compiler.cppstd:
+            check_min_cppstd(self, self._min_cppstd)
+        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
+        if minimum_version and Version(self.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(
+                f"{self.ref} requires C++{self._min_cppstd}, which your compiler does not support."
+            )
+
+        if self.options.shared:
+            if self._is_windows:
+                raise ConanInvalidConfiguration("Shared builds are currently not supported on Windows")
+            if os.getenv("CONAN_CENTER_BUILD_SERVICE") and self.settings.build_type == "Debug":
+                raise ConanInvalidConfiguration("Shared Debug build is not supported on CCI due to resource limitations")
+            if is_apple_os(self):
+                # FIXME iconv contains duplicate symbols in the libiconv and libcharset libraries (both of which are
+                #  provided by libiconv). This may be an issue with how conan packages libiconv
+                iconv_dep = self.dependencies.get("libiconv")
+                if iconv_dep and not iconv_dep.options.shared:
+                    raise ConanInvalidConfiguration("Static iconv cannot be linked into a shared library on macos "
+                                                    "due to duplicate symbols. Use libxml2/*:iconv=False.")
+
         if self.options.exceptions and not self.options.rtti:
-            message = 'Cannot enable exceptions without rtti support'
-            raise ConanInvalidConfiguration(message)
-        self._supports_compiler()
-        if cross_building(self, skip_x64_x86=True):
-            raise ConanInvalidConfiguration('Cross-building not implemented')
+            raise ConanInvalidConfiguration("Cannot enable exceptions without rtti support")
 
-    def source(self):
-        get(self, **self.conan_data["sources"][self.version], strip_root=True,
-            destination=self._source_subfolder)
-        self._patch_sources()
+        if cross_building(self):
```

Comment:
Should it not be `can_run` instead?

```suggestion
        if can_run(self):
```

Reply:
No - I specifically need to check for cross building (even if the host can run foreign architectures like Rosetta on Macos) as this needs special setup to build LLVM. This is left as an exercise for the next contributor!

Reply:
In the latter case, LLVM will attempt to build a native toolchain in a subdirector of the build directory. For this to work, that subdirectory would need to have the cmake configuration files generated by conan for the **build** platform - as well as the files generated for the **host** platform for the cross compilation build. The error is as follows - 

```
Configuring NATIVE LLVM...
FAILED: NATIVE/CMakeCache.txt 
cd /Users/jenkins/workspace/prod-v2/bsr/82122/cabdf/p/b/llvm-2a22810816590/b/build/Release/NATIVE && /Users/jenkins/cmake/cmake-3.20.1/bin/cmake -G Ninja -DCMAKE_MAKE_PROGRAM="/Users/jenkins/workspace/prod-v2/bsr/82122/cabdf/p/ninja7130a04140d30/p/bin/ninja" -DCMAKE_C_COMPILER=/Applications/conan/xcode/13.0/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DCMAKE_CXX_COMPILER=/Applications/conan/xcode/13.0/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ /Users/jenkins/workspace/prod-v2/bsr/82122/cabdf/p/b/llvm-2a22810816590/b/src -DLLVM_TARGET_IS_CROSSCOMPILE_HOST=TRUE -DLLVM_TARGETS_TO_BUILD="AArch64;AMDGPU;ARM;AVR;BPF;Hexagon;Lanai;Mips;MSP430;NVPTX;PowerPC;RISCV;Sparc;SystemZ;WebAssembly;X86;XCore" -DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD="" -DLLVM_DEFAULT_TARGET_TRIPLE="arm64-apple-darwin23.4.0" -DLLVM_TARGET_ARCH="host" -DLLVM_ENABLE_PROJECTS="" -DLLVM_EXTERNAL_PROJECTS="" -DLLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN="OFF" -DCMAKE_BUILD_TYPE=Release
```

Reply:
Thanks @planetmarshall  - this makes perfect sense. 

please feel free to add a link to https://github.com/conan-io/conan-center-index/pull/22997/files#r1740847096 in the recipe in the comments, so that in the future we know what the reason  in case it comes up